### PR TITLE
Convert ref and runtime packs to use the SharedFramework SDK and Arcade's installer tooling

### LINKED ISSUE #48013: Investigate using Microsoft.DotNet.SharedFramework.Sdk
https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.SharedFramework.Sdk

Runtime & WindowsDesktop use this to generate their SharedFx packages/Installers, but we have custom logic. We should investigate what it would take/what are the benefits of using this shared infra.

### LINKED ISSUE #49486: Microsoft.AspNetCore.App.runtimeconfig.json specifies runtime properties which values should not bet set by a shared framework
This is how the `Microsoft.AspNetCore.App.runtimeconfig.json` file looks like in a recent P7 build:

```JSON
{
  "runtimeOptions": {
    "tfm": "net8.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "8.0.0-preview.7.23359.3"
    },
    "rollForward": "LatestPatch",
    "configProperties": {
      "System.Runtime.Loader.UseRidGraph": true,
      "System.Reflection.Metadata.MetadataUpdater.IsSupported": false,
      "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization": false
    }
  }
}
```

The `configProperties` section should not be there. They got there via the process used to generate the shared framework. ASP.NET builds what looks like a framework dependent console app and uses the output to construct the shared framework. The `MetadataUpdater` and `BinaryFormatSerializer` properties are defaults set by the SDK for console apps. They may differ for other types of apps and ASP.NET shared framework should not dictate the values on its own, it's the job of the SDK to specify the defaults (and let the user override those).
The `UseRidGraph` is even worse, that one comes from the workaround used in the ASP.NET repo itself introduced here: https://github.com/dotnet/aspnetcore/pull/48908#issuecomment-1601894643. It must not be there when we ship.

Both runtime and WindowsDesktop had a similar issue for a long time now, but that has been fixed in https://github.com/dotnet/arcade/pull/13844. Unfortunately ASP.NET doesn't use Arcade to build the shared framework (unlike runtime and WindowsDesktop) - see https://github.com/dotnet/aspnetcore/issues/48013.

We need to fix at least the `UseRidGraph` before we ship, but ideally all of the properties should be removed.

/fyi @elinor-fung @agocke

### LINKED ISSUE #58073: Update aspnetcore to use new Arcade .deb creation tooling

### PR DESCRIPTION
**Pull request by jkoritzinsky**
# Convert ref and runtime packs to use the SharedFramework SDK and Arcade's installer tooling (for Linux installers)

Replace the custom infrastructure in ASP.NET Core for building shared frameworks, .debs, and .rpms with the Arcade infrastructure

## Description

Replace the custom infrastructure in this repo with the Arcade infrastructure. This provides a number of benefits:

- Crossgen2/R2R support is hooked up correctly through the SDK support
- Construction of targeting and runtime packs uses standard tooling, so less for aspnetcore to maintain and free updates to "best practices" from Arcade
- Support for building deb and rpm packages on any Linux distro

Fixes #48013
Fixes #49486
Fixes #58073
Contributes to https://github.com/dotnet/source-build/issues/3986

Depends on dotnet/arcade#15200
Depends on an SDK with dotnet/sdk#44436 (#58862)
~~Depends on dotnet/runtime#109169~~
Replaces #58600
Depends on https://github.com/dotnet/arcade/pull/15208
Depends on https://github.com/dotnet/arcade/pull/15209
Depends on https://github.com/dotnet/aspnetcore/pull/58934

### CODE REVIEW

**Comment by wtgodbe** [comment]
Whoa, awesome! Feel free to ping me on Teams if/when you want me to take a look

**Comment by jkoritzinsky** [comment]
I was going to handle messing with any Wix installers in a different PR (as I'm less confident in getting the upgrade codes and things right).

File: eng/Publishing.props [hunk 1]
```diff
@@ -31,13 +31,15 @@
     <_InstallersToPublish Include="$(ArtifactsDir)installers\**\*.version" UploadPathSegment="Runtime" Condition="'$(PublishInstallerBaseVersion)' == 'true'" />
 
     <!-- The following installers create checksums -->
-    <_InstallersToPublish Include="$(ArtifactsDir)installers\**\*.deb" UploadPathSegment="Runtime" ChecksumPath="%(FullPath).sha512" />
+    <_InstallersToPublish Include="$(ArtifactsDir)packages\**\*.deb" UploadPathSegment="Runtime" ChecksumPath="%(FullPath).sha512" />
```
Comment (wtgodbe) [inline eng/Publishing.props:34]: Is it the common pattern to publish installers to the `packages` dir?

File: global.json [hunk 2]
```diff
@@ -28,6 +28,9 @@
   },
   "msbuild-sdks": {
     "Microsoft.DotNet.Arcade.Sdk": "10.0.0-beta.24560.1",
-    "Microsoft.DotNet.Helix.Sdk": "10.0.0-beta.24560.1"
+    "Microsoft.DotNet.Helix.Sdk": "10.0.0-beta.24560.1",
+    "Microsoft.DotNet.SharedFramework.Sdk" : "10.0.0-beta.24560.1",
+    "Microsoft.Build.NoTargets": "3.7.0",
```
Comment (wtgodbe) [inline global.json:33]: Why these versions/when should we update them?

File: eng/Publishing.props (hunk 1 above)
Comment (jkoritzinsky) [reply eng/Publishing.props:34]: Yes, dotnet/runtime and dotnet/windowsdesktop publish all installers to the `packages` dir.

File: global.json (hunk 2 above)
Comment (jkoritzinsky) [reply global.json:33]: These are the versions used by dotnet/runtime today. I believe they're the newest versions. We generally try to update them when new versions come out, but we don't have a regular cadence.

File: src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        Overwrite="true" />
+  </Target>
+
+  <Target Name="PublishToSharedLayoutRoot" BeforeTargets="Build">
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
+             Targets="PublishToDisk"
+             Properties="OutputPath=$(TargetingPackLayoutRoot)" />
+  </Target>
+
+  <Target Name="AddLinuxPackageInformation" BeforeTargets="GetDebInstallerJsonProperties;GetRpmInstallerJsonProperties">
```
Comment (wtgodbe) [inline src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj:176]: Can you add a comment about what this target is for?

File: src/Framework/App.Runtime/src/aspnetcore-runtime-composite.proj [hunk 4]
```diff
@@ -0,0 +1,66 @@
+<Project Sdk="Microsoft.Build.Traversal">
+  <PropertyGroup>
```
Comment (wtgodbe) [inline src/Framework/App.Runtime/src/aspnetcore-runtime-composite.proj:2]: What's the purpose of this proj?

**Comment by wtgodbe** [comment]
Can you run an internal build of this branch so we can inspect the produced SharedFx/Targeting packs packages/installers?

File: src/Framework/App.Runtime/src/aspnetcore-runtime-composite.proj (hunk 4 above)
Comment (jkoritzinsky) [reply src/Framework/App.Runtime/src/aspnetcore-runtime-composite.proj:2]: This project builds an archive containing the .NET runtime and the ASP.NET Core runtime with the Composite ReadyToRun image (which is used in some of our docker images).

File: src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj (hunk 3 above)
Comment (jkoritzinsky) [reply src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj:176]: Will do

**Comment by jkoritzinsky** [comment]
Official build at https://dev.azure.com/dnceng/internal/_build/results?buildId=2580559&view=results

**Comment by ViktorHofer** [comment]
According to the official build, the targeting pack project is built and published from multiple legs:

```
##[error].packages\microsoft.dotnet.arcade.sdk\10.0.0-beta.24560.1\tools\SdkTasks\PublishBuildAssets.proj(42,5): error : Repeated Package entry: 'Microsoft.AspNetCore.App.Ref' - '10.0.0-alpha.2.24562.3' 
D:\a\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\10.0.0-beta.24560.1\tools\SdkTasks\PublishBuildAssets.proj(42,5): error : InvalidOperationException: Duplicate package entries are not allowed for publishing to BAR, as this can cause race conditions and unexpected behavior
D:\a\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\10.0.0-beta.24560.1\tools\SdkTasks\PublishBuildAssets.proj(42,5): error :    at Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.MergeManifests(List`1 parsedManifests) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line 557
D:\a\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\10.0.0-beta.24560.1\tools\SdkTasks\PublishBuildAssets.proj(42,5): error :    at Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.PushMetadataAsync(CancellationToken cancellationToken) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line 105
D:\a\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\10.0.0-beta.24560.1\tools\SdkTasks\PublishBuildAssets.proj(42,5): error : 

```

**Comment by jkoritzinsky** [comment]
Failures in the `Linux x64` leg will be fixed by #58934

**Review by wtgodbe — APPROVED** [review]
LGTM, the only other thing would be to confirm the manifest from an internal build of this branch matches the manifest from an internal build of main

**Comment by jkoritzinsky** [comment]
I ran an official build and it looks fine. However, the manually-triggered official build does in-build signing instead of post-build, so comparing apples-to-apples is difficult. I think this is okay for merge, but let me know if you think more validation is necessary.

**Comment by jkoritzinsky** [comment]
Fixed an issue with in-build signing, new official build at https://dev.azure.com/dnceng/internal/_build/results?buildId=2595419&view=results

**Comment by jkoritzinsky** [comment]
New official build link: https://dev.azure.com/dnceng/internal/_build/results?buildId=2595988&view=results

**Comment by jkoritzinsky** [comment]
New official build link: https://dev.azure.com/dnceng/internal/_build/results?buildId=2596133&view=results

**Comment by jkoritzinsky** [comment]
Okay there's an ongoing infra issue that broke the official build. If this build produces the right installers for Linux arm64, we're good to merge (as that's the last thing I needed to validate).

**Comment by jkoritzinsky** [comment]
ARM64 rpms are being produced, setting auto-merge.

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.sfxproj [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    <RuntimeIdentifier>$(TargetRuntimeIdentifier)</RuntimeIdentifier>
+    <!-- System.Diagnostics.EventLog.Messages.dll has no file version -->
+    <PermitDllAndExeFilesLackingFileVersion>true</PermitDllAndExeFilesLackingFileVersion>
+    <!-- Target the latest runtime patch so the latest version is specified in runtimeconfig.json -->
+    <TargetLatestRuntimePatch>true</TargetLatestRuntimePatch>
+    <RollForward>LatestPatch</RollForward>
+    <!-- Precompile the shared framework with ReadyToRun. ReadyToRun is not currently supported on s390x or ppc64le or armv6 or loongarch64. -->
+    <PublishReadyToRun Condition=" '$(TargetArchitecture)' == 's390x' OR '$(TargetArchitecture)' == 'armv6' OR '$(TargetArchitecture)' == 'ppc64le' OR '$(TargetArchitecture)' == 'loongarch64' ">false</PublishReadyToRun>
+    <!-- VMR doesn't produce valid crossgen for the host/build machine https://github.com/dotnet/source-build/issues/3793 -->
+    <PublishReadyToRun Condition=" '$(DotNetBuild)' == 'true' AND '$(Crossbuild)' == 'true' ">false</PublishReadyToRun>
```
Comment (ViktorHofer) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.sfxproj:19]: I was surprised to see this. Are you sure that this is still an issue?

Comment (jkoritzinsky) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.sfxproj:19]: Yes I believe that is still the case (to the point that I think I messed up this condition and we're seeing failures related to it in the Aspnet->SDK update).

I have some thoughts about it and I'll open an issue today to track fixing it.

**Pull request merged**